File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccessTesting/FakeSettingsPort.swift
import SecretaryVaultAccess

/// In-memory `SettingsPort` double with spies + failure injection, modeled on
/// `FakeTrashPort`. Seeds a settings value + bounds; `writeSettings` records the
/// captured value to `writtenSettings` and updates the seed (so a later read
/// reflects it). `failNextRead` / `failNextWrite` throw once then clear.
/// `@unchecked Sendable` for the same single-thread reason as `FakeTrashPort`.
public final class FakeSettingsPort: SettingsPort, @unchecked Sendable {
    public var settings: VaultSettings
    public var bounds: SettingsBounds
    public var failNextRead: VaultAccessError?
    public var failNextWrite: VaultAccessError?

    public private(set) var readCount = 0
    public private(set) var writtenSettings: [VaultSettings] = []

    public init(settings: VaultSettings = FakeSettingsPort.defaultSettings,
                bounds: SettingsBounds = FakeSettingsPort.defaultBounds) {
        self.settings = settings
        self.bounds = bounds
    }

    public func readSettings() throws -> VaultSettings {
        readCount += 1
        if let e = failNextRead { failNextRead = nil; throw e }
        return settings
    }

    public func writeSettings(_ settings: VaultSettings) throws {
        if let e = failNextWrite { failNextWrite = nil; throw e }   // throw before recording: no write happened
        writtenSettings.append(settings)
        self.settings = settings
    }

    public func settingsBounds() -> SettingsBounds { bounds }

    /// The bridge-schema defaults (auto-lock 10 min, require-password true,
    /// grace 2 min, retention 90 d) — what an absent settings block yields.
    public static let defaultSettings = VaultSettings(
        autoLockTimeoutMs: 600_000,
        requirePasswordBeforeEdits: true,
        reauthGraceWindowMs: 120_000,
        retentionWindowMs: 90 * 86_400_000)

    /// The real projected bound values (retention 1…3650 d default 90;
    /// grace 0…60 min default 2).
    public static let defaultBounds = SettingsBounds(
        retentionDefaultMs: 90 * 86_400_000,
        retentionMinMs: 86_400_000,
        retentionMaxMs: 3650 * 86_400_000,
        reauthGraceDefaultMs: 120_000,
        reauthGraceMinMs: 0,
        reauthGraceMaxMs: 3_600_000)
}

File: android/vault-access/src/test/kotlin/org/secretary/browse/FakeSettingsPort.kt
package org.secretary.browse

import kotlinx.coroutines.CompletableDeferred

/**
 * In-memory [SettingsPort] test double. Seed [settings] / [bounds]; prime [failNextRead] /
 * [failNextWrite] to throw once (then they self-clear). Records each write in [writtenSettings].
 * Mirror of iOS `FakeSettingsPort`.
 *
 * When [writeGate] is set, [writeSettings] awaits it before recording — lets a test hold a write in
 * flight while a second call hits the model's re-entrancy guard (a faithful race, not a sleep).
 */
class FakeSettingsPort(
    var settings: VaultSettings = defaultVaultSettings(),
    private val bounds: SettingsBounds = defaultSettingsBounds(),
    var failNextRead: VaultBrowseError? = null,
    var failNextWrite: VaultBrowseError? = null,
    private val writeGate: CompletableDeferred<Unit>? = null,
) : SettingsPort {
    val writtenSettings = mutableListOf<VaultSettings>()

    override fun readSettings(): VaultSettings {
        failNextRead?.let { failNextRead = null; throw it }
        return settings
    }

    override suspend fun writeSettings(settings: VaultSettings) {
        writeGate?.await()
        failNextWrite?.let { failNextWrite = null; throw it }
        writtenSettings += settings
        this.settings = settings
    }

    override fun settingsBounds(): SettingsBounds = bounds
}

/** The real schema defaults, so tests seed the same values production sees for a fresh vault
 *  (bridge `Settings::default()`: auto-lock 10 min, require-password on, grace 2 min, retention 90 d). */
fun defaultVaultSettings(): VaultSettings = VaultSettings(
    autoLockTimeoutMs = 600_000L,
    requirePasswordBeforeEdits = true,
    reauthGraceWindowMs = 120_000L,
    retentionWindowMs = 90L * MS_PER_DAY,
)

/** The real projected bound constants (the six uniffi reader values). */
fun defaultSettingsBounds(): SettingsBounds = SettingsBounds(
    retentionDefaultMs = 90L * MS_PER_DAY,
    retentionMinMs = MS_PER_DAY,
    retentionMaxMs = 3650L * MS_PER_DAY,
    reauthGraceDefaultMs = 120_000L,
    reauthGraceMinMs = 0L,
    reauthGraceMaxMs = 3_600_000L,
)
